### PR TITLE
stm32f1: Fix SPI issues

### PR DESCRIPTION
**Pull request by yschroeder**
The SPI implementation of the STM32F1 reads wrong bytes when the previous transfer did not read bytes at all but only sent data.

This PR ensures that the receive buffer is empty in case it will be read.

Before, the first byte of an SPI transfer could be the last byte received by the previous transfer (in case it did not read the receive buffer).

### CODE REVIEW

**Comment by grigorig** [comment]
Shouldn't it be sufficient to simply make this a single unconditional read? I think the SPI implementation only has a one-byte buffer.

**Comment by yschroeder** [comment]
Yes, I was thinking about refactoring the `spi_transfer()` overall to be closer to the one used on AVR. I will update the PR with a cleaner `spi_transfer()` implementation.

**Comment by yschroeder** [comment]
@grigorig: Should I clean this up or do you want to take care of this together with #909?

**Comment by yschroeder** [comment]
I reworked the code that fixes the SPI receive problem as discussed above.

Further I went ahead and tackled issue #909. I reworked a lot of the SPI code. The old code directly worked on the registers of `SPI2`. As I understand Klipper, it should be possible to have different SPI configurations for different devices attached to the MCU. The new code initializes a `LL_SPI_InitTypeDef` struct for each device configured via `spi_config()` and ensures that this config is loaded in `spi_prepare()`.

I tested it with a TMC2130 and observed the SPI transfers on a logic analyzer. The transfers are correct, the stepper driver is configured, the output of `DUMP_TMC` looks valid and the stepper motor is turning.

As I am no expert on STM32 MCUs, someone with more experience should review the code.

File: src/stm32f1/Makefile [hunk 1]
```diff
@@ -11,6 +11,7 @@ CFLAGS += -mthumb -mcpu=cortex-m3
 CFLAGS += -Ilib/cmsis-core
 CFLAGS += -Ilib/cmsis-stm32f1/include -Ilib/hal-stm32f1/include
 CFLAGS += -DSTM32F103xB
+CFLAGS += -DUSE_FULL_LL_DRIVER
```
Comment (yschroeder) [inline src/stm32f1/Makefile]: Needed to use `SPI_InitTypeDef`.

File: src/stm32f1/main.c [hunk 2]
```diff
@@ -112,8 +112,6 @@ void adc_config(void)
 void spi_config(void)
 {
     LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_SPI2);
-    LL_SPI_SetNSSMode(SPI2, LL_SPI_NSS_SOFT);
-    LL_SPI_SetMode(SPI2, LL_SPI_MODE_MASTER);
```
Comment (yschroeder) [inline src/stm32f1/main.c]: These are not needed anymore as they are configured in every `spi_prepare()`.

File: src/stm32f1/spi.c [hunk 3]
```diff
@@ -70,23 +70,30 @@ static void spi_init_pins(void)
...
+    spi_init_pins();
 
     return config;
 }
 
 void
 spi_prepare(struct spi_config config)
 {
-    *SPI2 = config.config;
+    LL_SPI_Init(SPI2, &config.config);
```
Comment (yschroeder) [inline src/stm32f1/spi.c]: This configures the SPI for each transfer individually.

File: src/stm32f1/spi.c [hunk 4]
```diff
@@ -97,9 +104,10 @@ spi_transfer(struct spi_config config, uint8_t receive_data,
     while (len--) {
         LL_SPI_TransmitData8(SPI2, *data);
         while (!LL_SPI_IsActiveFlag_TXE(SPI2));
+        while (!LL_SPI_IsActiveFlag_RXNE(SPI2));
+        uint8_t rdata = LL_SPI_ReceiveData8(SPI2);
```
Comment (yschroeder) [inline src/stm32f1/spi.c]: Always read the incoming SPI data, do nothing with it if a read was not requested.

**Comment by yschroeder** [comment]
Ping @KevinOConnor

Do you have any remarks on this code that stop it from being merged? There are a few people in the issue section that need this fix and I think it would be nice for the next release to have this breaking bug of the STM32F1 platform fixed.

**Comment by grigorig** [comment]
LGTM, but I didn't test it. Sorry, I still haven't found time myself to look at this issue. Too much going on!

**Comment by KevinOConnor** [comment]
I've pulled in the first commit - thanks!  I'm going to hold off on the second commit, because it's unclear to me what the implication of "-DUSE_FULL_LL_DRIVER" is.  I understand that the STM32F103 SPI code does not work right now, but as near as I can tell it never worked.  My main focus for the next week is no regressions.  So, this will wait until after the Klipper v0.7.0 release.

-Kevin

**Comment by yschroeder** [comment]
I can understand that and I also do not know what the implications are. I looked what it does for SPI and in the header it only defines the extra `LL_SPI_InitTypeDef`. I don't know though what it does for the other components.

FWIW, I have steppers running without problems, so it works for me currently and I can continue my work on the cable driven robot/printer.

**Comment by jakep82** [comment]
If it's of any benefit, I can test this with a MAX31865 module this weekend.  I have a blue pill on a breadboard for testing with Klipper.

**Comment by yschroeder** [comment]
I researched about the use of USE_FULL_LL_DRIVER. I found the documentation on the CubeMX output from ST:

https://www.st.com/resource/en/user_manual/dm00151047.pdf

I learned the following: The drivers by ST consist of two components. The HAL and the LL part. The LL part implements low level drivers with less overhead, while the HAL offers higher level easy to use drivers. There are multiple ways to use the drivers. You can only use the HAL (which itself uses the LL drivers) or you can use the HAL to initialize the hardware registers and use the LL to use the hardware with less overhead.

The third option is to only use the LL driver (which is what we do). If you want to do this, you need to supply the USE_FULL_LL_DRIVER definition to enable the LL initialization code. This seems to be a safety switch to avoid unintended use of the LL code by users who actually want to use the HAL.

So basically, we were missing the USE_FULL_LL_DRIVER in the first place.

**Comment by KevinOConnor** [comment]
I was able to take a look at this and replicate the problem on my "blue pill" device with an ssd1306 spi display.

I think the stm32 library abstractions are making this more complicated than it actually is.  There's just one register (SPI_CR1) that needs to be set in spi_prepare().  The patch below gets the interface working for me.

-Kevin

```diff
diff --git a/src/stm32f1/gpio.h b/src/stm32f1/gpio.h
index a4a3c121..8aba2f85 100644
--- a/src/stm32f1/gpio.h
+++ b/src/stm32f1/gpio.h
@@ -33,7 +33,7 @@ uint16_t gpio_adc_read(struct gpio_adc g);
 void gpio_adc_cancel_sample(struct gpio_adc g);
 
 struct spi_config {
-    SPI_TypeDef config;
+    uint32_t spi_cr1;
 };
 struct spi_config spi_setup(uint32_t bus, uint8_t mode, uint32_t rate);
 void spi_prepare(struct spi_config config);
diff --git a/src/stm32f1/main.c b/src/stm32f1/main.c
index b91d4dd0..e7a0cd91 100644
--- a/src/stm32f1/main.c
+++ b/src/stm32f1/main.c
@@ -112,8 +112,6 @@ void adc_config(void)
 void spi_config(void)
 {
     LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_SPI2);
-    LL_SPI_SetNSSMode(SPI2, LL_SPI_NSS_SOFT);
-    LL_SPI_SetMode(SPI2, LL_SPI_MODE_MASTER);
 }
 
 void io_config(void)
diff --git a/src/stm32f1/spi.c b/src/stm32f1/spi.c
index a90a030a..5ce81fb9 100644
--- a/src/stm32f1/spi.c
+++ b/src/stm32f1/spi.c
@@ -70,24 +70,24 @@ static void spi_init_pins(void)
 struct spi_config
 spi_setup(uint32_t bus, uint8_t mode, uint32_t rate)
 {
-    struct spi_config config;
-    config.config = *SPI2;
-
     if (bus > 0 || !rate)
         shutdown("Invalid spi_setup parameters");
 
     spi_init_pins();
-    spi_set_mode(&config.config, mode);
-    spi_set_baudrate(&config.config, rate);
+    SPI_TypeDef spi_hw = { };
+    LL_SPI_SetNSSMode(&spi_hw, LL_SPI_NSS_SOFT);
+    LL_SPI_SetMode(&spi_hw, LL_SPI_MODE_MASTER);
+    spi_set_mode(&spi_hw, mode);
+    spi_set_baudrate(&spi_hw, rate);
+    LL_SPI_Enable(&spi_hw);
 
-    return config;
+    return (struct spi_config){ .spi_cr1 = spi_hw.CR1 };
 }
 
 void
 spi_prepare(struct spi_config config)
 {
-    *SPI2 = config.config;
-    LL_SPI_Enable(SPI2);
+    SPI2->CR1 = config.spi_cr1;
 }
 
 void
```

**Comment by KevinOConnor** [comment]
Any further comments on this?

-Kevin

**Comment by yschroeder** [comment]
Sorry, I did not find time yet to test this. I'm not sure if what you implemented is the intended usage of that low level SPI driver from ST. If it works, I'm fine with doing it your way.

**Comment by KevinOConnor** [comment]
Okay, lets close this for now then.  The master code is working in my tests.

-Kevin
